clang/lib/CodeGen/CodeGenFunction.cpp

```diff
@@ -2854,10 +2854,121 @@ void CodeGenFunction::EmitMultiVersionResolver(
   case llvm::Triple::aarch64:
     EmitAArch64MultiVersionResolver(Resolver, Options);
     return;
+  case llvm::Triple::riscv32:
+  case llvm::Triple::riscv64:
+    EmitRISCVMultiVersionResolver(Resolver, Options);
+    return;
 
   default:
-    assert(false && "Only implemented for x86 and AArch64 targets");
+    assert(false && "Only implemented for x86, AArch64 and RISC-V targets");
+  }
+}
+
+void CodeGenFunction::EmitRISCVMultiVersionResolver(
+    llvm::Function *Resolver, ArrayRef<MultiVersionResolverOption> Options) {
+
+  if (getContext().getTargetInfo().getTriple().getOS() !=
+      llvm::Triple::OSType::Linux) {
+    CGM.getDiags().Report(diag::err_os_unsupport_riscv_target_clones);
+    return;
+  }
+
+  llvm::BasicBlock *CurBlock = createBasicBlock("resolver_entry", Resolver);
+  Builder.SetInsertPoint(CurBlock);
+  EmitRISCVCpuInit();
+
+  bool SupportsIFunc = getContext().getTargetInfo().supportsIFunc();
+  bool HasDefault = false;
+  unsigned DefaultIndex = 0;
+  // Check the each candidate function.
+  for (unsigned Index = 0; Index < Options.size(); Index++) {
```

Comment:
There's a semantic problem here unless I'm missing something.   Say the user specifies three choices: default, zba, and zba+zbb (in that order.)  Unless I'm missing something, the third case would be dead here which is almost certainly not what the user intended.  How do other targets handle this case?

Reply:
For RISC-V target, I'm not sure that it is possible to define the score for all extension combination, then we could sort the order base on it. Or we could extend the `target_clones` syntax with `Priority=<NUM>` to indicate the order by user. 

```
__attribute__((target_clones("default", "arch=+zba;Priority=1", "arch=+zba,+zbb;Priority=2"))) int bar() {  
    return 1;
}
```

And Yes, It will third version (+zba,+zbb) will be dead in current implementation. It will always get into +zba version or default.

Reply:
I think this question needs some focused discussion.

My initial stab at this would be something along the following:
* Reorder any super set of extensions before a subset.
* Respect user provided order otherwise.

From the precedent you point to, we can reasonably treat the ordering an implementation defined heuristic.  We should document that.

Reply:
> I think this question needs some focused discussion.
> 
> My initial stab at this would be something along the following:
> 
> * Reorder any super set of extensions before a subset.
> * Respect user provided order otherwise.
> 
> From the precedent you point to, we can reasonably treat the ordering an implementation defined heuristic. We should document that.

Hi @preames, Here is another [spec update](https://github.com/riscv-non-isa/riscv-c-api-doc/pull/85) to address version ordering issue, and it be implemented inside this patch.